cardano-crypto-class/src/Cardano/Crypto/DSIGN/EcdsaSecp256k1.hs

```diff
@@ -55,9 +81,198 @@ import Cardano.Crypto.DSIGN.Class (
+          -- maximum permissible length is given as a pointer, which is
+          -- overwritten to indicate the number of bytes we actually wrote; if
+          -- we get a mismatch, then the serialization failed. While an odd
+          -- choice, we have to go with it.
+          alloca $ \(lenPtr :: Ptr CSize) -> do
+            poke lenPtr len
+            void $ secpEcPubkeySerialize secpCtxPtr dstp lenPtr psp secpEcCompressed
+            writtenLen <- peek lenPtr
+            unless (writtenLen == len) (error "rawSerializeVerKeyDSIGN: Did not write correct length")
+    rawSerialiseSignKeyDSIGN (SignKeyEcdsaSecp256k1 psb) = psbToByteString psb
+    {-# NOINLINE rawDeserialiseSigDSIGN #-}
+    rawDeserialiseSigDSIGN bs = 
+      SigEcdsaSecp256k1 <$> (psbFromByteStringCheck bs >>= go)
+      where
+        go :: 
+          PinnedSizedBytes SECP256K1_ECDSA_SIGNATURE_BYTES -> 
+          Maybe (PinnedSizedBytes SECP256K1_ECDSA_SIGNATURE_BYTES_INTERNAL)
+        go psb = unsafeDupablePerformIO . psbUseAsSizedPtr psb $ \psp -> do
+          -- This is a slightly odd order of operations, but it's forced on us,
+          -- as psbCreateSized only allows an initialization that returns the
+          -- trivial value. Instead, we allocate (but do not initialize) a
+          -- PinnedSizedBytes of the right size, then try using it. This is
+          -- safe, as if deserialization fails, then we never get to see the
+          -- PinnedSizedBytes this would produce (ostensibly full of junk), and
+          -- if deserialization succeeds, it'll overwrite the contents anyway.
+          sigPsb <- psbCreateSized (\_ -> pure ())
+          res <- psbUseAsSizedPtr sigPsb $ \sigp -> 
+                  secpEcdsaSignatureParseCompact secpCtxPtr sigp psp
+          pure $ case res of 
+            1 -> pure sigPsb
+            _ -> Nothing
+    {-# NOINLINE rawDeserialiseVerKeyDSIGN #-}
+    rawDeserialiseVerKeyDSIGN bs = 
+      VerKeyEcdsaSecp256k1 <$> (psbFromByteStringCheck bs >>= go)
+      where
+        -- In addition to the same weirdness as above, here we also have to
+        -- 'forget' our size temporarily, as the underlying C API allows
+        -- compression to _any_ size, not just our particular one. This is still
+        -- safe, but it requires 'unwrapping' a SizedPtr.
+        go :: 
+          PinnedSizedBytes SECP256K1_ECDSA_PUBKEY_BYTES -> 
+          Maybe (PinnedSizedBytes SECP256K1_ECDSA_PUBKEY_BYTES_INTERNAL)
+        go psb = unsafeDupablePerformIO . psbUseAsSizedPtr psb $ \(SizedPtr p) -> do
+          let srcp = castPtr p
+          let srcLen :: CSize = fromIntegral . natVal $ Proxy @SECP256K1_ECDSA_PUBKEY_BYTES
+          vkPsb <- psbCreateSized (\_ -> pure ())
+          res <- psbUseAsSizedPtr vkPsb $ \vkp -> 
+                  secpEcPubkeyParse secpCtxPtr vkp srcp srcLen
+          pure $ case res of 
+            1 -> pure vkPsb
+            _ -> Nothing
```

Comment:
This will avoid unnecessary ByteArray casting and will read a bit cleaner
```suggestion
          (vkPsb, res) <- psbCreateSizedReturn $ \(SizedPtr ptr) -> do
            let vkp = castPtr :: Ptr CUChar
            secpEcPubkeyParse secpCtxPtr vkp srcp srcLen
          pure $ case res of 
            1 -> Just vkPsb
            _ -> Nothing
```